mqtt-v5-broker/src/plugin.rs

```diff
@@ -0,0 +1,124 @@
+use log::{trace, warn};
+use mqtt_v5::types::{
+    AuthenticatePacket, ConnectPacket, ConnectReason, PublishAckPacket, PublishAckReason,
+    PublishPacket, PublishReceivedPacket, PublishReceivedReason, QoS, SubscribeAckPacket,
+    SubscribeAckReason, SubscribePacket,
+};
+
+pub struct Noop;
+
+/// Result of a authentication attempt
+pub enum AuthentificationResult {
+    /// Authentification reason
+    Reason(ConnectReason),
+    /// Send this auth packet to the client and wait for the response.
+    Packet(AuthenticatePacket),
+}
+
+/// Broker plugin
+pub trait Plugin {
+    /// Called on connect packet reception
+    fn on_connect(&mut self, packet: &ConnectPacket) -> AuthentificationResult;
+
+    /// Called on authenticate packet reception
+    fn on_authenticate(&mut self, packet: &AuthenticatePacket) -> AuthentificationResult;
+
+    /// Called on subscribe packets reception
+    fn on_subscribe(&mut self, packet: &SubscribePacket) -> SubscribeAckPacket;
+
+    /// Called on publish packets reception for QoS 0. Return true if the packet should be published to the clients.
+    fn on_publish_received_qos0(&mut self, packet: &PublishPacket) -> bool;
```

Comment:
What's the reasoning for splitting out publishes based on the qos? `PublishPacket` already has the `qos` field on it, but maybe there's some other reasoning for it that I'm missing.

Is this for handling the individual steps in the QoS sending handshake, or just a wholesale "We received a qos2 packet, do you want to send it?"

Reply:
I had a single "callback" for publication but struggled with the return type. There's no return for QOS0, a Publish Ack for QOS1 and PublishReceivedAck for QOS2. 
I'm fine with merging those if you want and but would require to define a struct for that the includes basically the same as the `PublishAckPacket` and `PublishReceivedPacket` in an `Option`.